Added shared drive crawler

Pourliver opened the pull request:
I added the automatic shared drive crawler I had in mind, as shared a little in #137.

This has been done after #140, so its been easier to implement since I knew the download flow and quirks.

I had to create a crawler_config folder since I didn't like the idea of mixing classes definition and plaintext files.

I tried to put more documentation in this class since the core logic may be hard to follow. I've tested it a lot and it works amazingly well, but please do test it.

Res260 commented:
Idea: log in a file each file, even if not downloaded

Res260 commented on pyrdp/mitm/FileCrawler.py:
```diff
@@ -0,0 +1,277 @@
+            Path(localPath).parent.mkdir(parents=True, exist_ok=True)
+            targetFile = open(localPath, "wb")
+        except Exception as e:
+            self.log.error("Cannot save file: %(exception)s", {"exception": str(e)})
```
`log.exception` would be better

Res260 commented on pyrdp/mitm/FileCrawler.py:
```diff
@@ -0,0 +1,277 @@
+        self.downloadFiles[remotePath] = targetFile
+        self.deviceRedirection.sendForgedFileRead(file.deviceID, remotePath)
+
+    def listDirectory(self, deviceID: int, path: str, download: bool = False):
```
docstring would help

Res260 commented on pyrdp/mitm/config.py:
```diff
@@ -53,6 +53,15 @@ def __init__(self):
         """Amount of time the payload should take to complete, in milliseconds"""
 
+        self.disableCrawler: bool = False
+        """Whether the crawler should be crawling or not"""
```
should be disabled or not*, to match the boolean value

Res260 commented on pyrdp/mitm/crawler_config/match.txt:
```diff
@@ -0,0 +1,25 @@
+*/*.bat
+*/*.exe
+*/*.dll
+*/*.ps1
```
.py? possibly other languages as well

Pourliver replied:
Do you have any input on this? @obilodeau This file is what we currently download from the disk, folder AND files.

Pourliver replied:
I think that crawling code files would give us a lot of thrash and would expand the crawl time. I went with compiled files and scripts that work out-of-the-box on Windows / Linux.

Res260 commented on bin/pyrdp-mitm.py:
```diff
@@ -164,6 +164,9 @@ def main():
     parser.add_argument("--payload-duration", help="Amount of time for which input / output should be dropped, in milliseconds. This can be used to hide the payload screen.", default=None)
+    parser.add_argument("--disable-crawler", help="Disable automatic shared drive scraping", action="store_true")
+    parser.add_argument("--crawler-match-file", help="File to be used by the crawler to chose what to download when scraping the client shared drives.")
+    parser.add_argument("--crawler-ignore-file", help="File to be used by the crawler to chose what folders to avoid when scraping the client shared drives.")
```
put default value?

Pourliver replied:
Since the configuration takes places in `bin/pyrdp-mitm.py` and the default file is in `pyrdp/mitm/crawler_config/match.txt`, I don't think moving the default value in the configuration parser is a good idea. Don't forget that the `bin/` content gets moved to the user binaries folder during installation, and we lose direct path references.

Res260 replied:
makes sense

Pourliver commented:
> Idea: log in a file each file, even if not downloaded

To add to this idea, since we talked about it in person, the goal is to keep track of what the crawler is seeing, so we can further improve the default match / ignore regexes later.

Pourliver commented on pyrdp/mitm/FileCrawler.py:
```diff
@@ -0,0 +1,277 @@
+        self.path = filePath
+        self.isDirectory = isDirectory
+
+class FileCrawler(DeviceRedirectionMITMObserver):
```
I'm not too sure the value that it brings, but I can rename it

Res260 replied:
It was for consistence and verboseness, but I dont think we're constant in the rest of the code base so doesnt matter for now

Pourliver commented:
The PR is ready for another review. I added a new logger that is not from the root logger, so we could log in another file without messing with STDOUT. This new file keeps track of every file we crawled.

Res260 commented:
Oops commented on the commet instead of the PR:

Besides the exception() log TODO, LGTM!

Pourliver commented:
Sorry, but I'm not too sure I understand whats wrong @Res260 , could you please elaborate a bit?

Res260 commented:
I think I saw you added a log.exception() somewhere else in the code, so I
thought it would be easy to implement instead of leaving a TODO in the code.

On Wed., Aug. 14, 2019, 13:16 Maxime Carbonneau, <notifications@github.com>
wrote:

> Sorry, but I'm not too sure I understand whats wrong @Res260
> <https://github.com/Res260> , could you please elaborate a bit?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoSecure/pyrdp/pull/141?email_source=notifications&email_token=ADPMNL3QRDF5KDJYIF6FRYLQEQ4WPA5CNFSM4IITPNLKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4JPPEQ#issuecomment-521336722>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADPMNLYIG64XYKMPOBRMKRTQEQ4WPANCNFSM4IITPNLA>
> .
>

Pourliver commented:
You're right, I totally missed this one, my bad! Fixed in latest commit.

Res260 commented:
Lgtm

Pourliver commented:
Fixed a small problem with the parser where the last line of each file got truncated, and simplified the match file since `*/*.doc*` can also match `*/*.docker`, which is unintended.

xshill commented on bin/pyrdp-mitm.py:
```diff
@@ -169,6 +191,9 @@ def main():
     parser.add_argument("--payload-powershell-file", help="PowerShell script to run automatically upon connection (as -EncodedCommand)", default=None)
     parser.add_argument("--payload-delay", help="Time to wait after a new connection before sending the payload, in milliseconds", default=None)
     parser.add_argument("--payload-duration", help="Amount of time for which input / output should be dropped, in milliseconds. This can be used to hide the payload screen.", default=None)
+    parser.add_argument("--disable-crawler", help="Disable automatic shared drive scraping", action="store_true")
```
I think this is such a big feature that it should be enabled consciously, so I'll change this to opt-in.
